merkle/src/merkle.rs

```diff
@@ -37,21 +39,204 @@ use std::ops;
 ///
 /// TODO: Ord
 #[derive(Debug, Clone, Eq, PartialEq)]
-pub struct MerkleTree<T: Ord + Clone + AsRef<[u8]> + Sync + Send, A: Algorithm<T>> {
-    data: Vec<T>,
+pub struct MerkleTree<T, A, K>
+where
+    T: Element,
+    A: Algorithm<T>,
+    K: Store<T>,
+{
+    data: K,
     leafs: usize,
     height: usize,
     _a: PhantomData<A>,
+    _t: PhantomData<T>,
+}
+
+/// Element stored in the merkle tree.
+pub trait Element: Ord + Clone + AsRef<[u8]> + Sync + Send + Default + std::fmt::Debug {
+    /// Returns the length of an element when serialized as a byte slice.
+    fn byte_len() -> usize;
+
+    /// Creates the element from its byte form. Panics if the slice is not appropriately sized.
+    fn from_slice(bytes: &[u8]) -> Self;
+}
+
+/// Backing store of the merkle tree.
+pub trait Store<E: Element>: ops::Deref<Target = [E]> + std::fmt::Debug + Clone {
+    /// Creates a new store which can store up to `size` elements.
+    fn new(size: usize) -> Self;
+
+    fn new_from_slice(size: usize, data: &[u8]) -> Self;
+
+    fn write_at(&mut self, el: E, i: usize);
+
+    fn read_at(&self, i: usize) -> E;
+    fn read_range(&self, r: ops::Range<usize>) -> Vec<E>;
+
+    fn len(&self) -> usize;
+    fn is_empty(&self) -> bool;
+    fn push(&mut self, el: E);
+}
+
+#[derive(Debug, Clone)]
+pub struct VecStore<E: Element>(Vec<E>);
+
+impl<E: Element> ops::Deref for VecStore<E> {
+    type Target = [E];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<E: Element> Store<E> for VecStore<E> {
+    fn new(size: usize) -> Self {
+        VecStore(Vec::with_capacity(size))
+    }
+
+    fn write_at(&mut self, el: E, i: usize) {
+        if self.0.len() <= i {
+            self.0.resize(i + 1, E::default());
+        }
+
+        self.0[i] = el;
+    }
+
+    fn new_from_slice(size: usize, data: &[u8]) -> Self {
+        let mut v: Vec<_> = data
+            .chunks_exact(E::byte_len())
+            .map(E::from_slice)
+            .collect();
+        let additional = size - v.len();
+        v.reserve(additional);
+
+        VecStore(v)
+    }
+
+    fn read_at(&self, i: usize) -> E {
+        self.0[i].clone()
+    }
+
+    fn read_range(&self, r: ops::Range<usize>) -> Vec<E> {
+        self.0.index(r).to_vec()
+    }
+
+    fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn push(&mut self, el: E) {
+        self.0.push(el);
+    }
+}
+
+#[derive(Debug)]
+pub struct MmapStore<E: Element> {
+    store: MmapMut,
+    len: usize,
+    _e: PhantomData<E>,
+}
+
+impl<E: Element> ops::Deref for MmapStore<E> {
+    type Target = [E];
+
+    fn deref(&self) -> &Self::Target {
+        unimplemented!()
```

Comment:
@dignifiedquire This is conversion that stopped me short initially and I need to resolve it to test the `mmap` version, how should this be implemented? I've noticed that you tried to avoid `mem::transmute` which I originally thought of in https://github.com/filecoin-project/merkle_light/issues/8#issuecomment-468829569, but I can't think of any alternatives at the moment, WDYT?

Reply:
what is the code path that relies on this? I am not sure we can actually implement it correctly

Reply:
https://circleci.com/gh/filecoin-project/rust-fil-proofs/15226?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Reply:
can you try running these with `RUST_BACKTRACE=1` so we can see where this originates from?

Reply:
https://github.com/filecoin-project/rust-fil-proofs/blob/master/storage-proofs/src/drgporep.rs#L273

and related.

Just so I follow your thought process in the design, was the `unimplemented` macro left there intentionally? If so, why was the deref implemented for the rest of the structures? (e.g., `MerkleTree`, `VecStore`) Should we avoid using `transmute`?

Reply:
- yes, we should avoid transmute if at all possible (it leads down a dark path of unsafety)
- I left it in there to be able to reuse existing tests, but otherwise I think that we should try to avoid relying on deref if we can

Reply:
it should be easy to just use the read methods for the linked code instead

Reply:
hmm, maybe not that straight forward, this is tricky

Reply:
we could easily support `AsRef<[u8]>` without being unsafe, maybe that could help us in that case?

Reply:
> we could easily support `AsRef<[u8]>` without being unsafe, maybe that could help us in that case?

Would that be for the entire tree or each element independently? (the `AsRef<[u8]>` support)

Reply:
I meant for the entire tree

Reply:
I think that may work, but how do we implement something like that for the entire tree without unsafe things like `transmute`? That was what blocked me in the first place, the `[T] <-> [u8]` conversion (as originally described in https://github.com/filecoin-project/merkle_light/issues/8).

Reply:
well the mmap already gives you `&[u8]` for free, so the hard part in this case is getting it done for the vector, which....hmmmm is hard to say the least as even a transmute would give you the wrong thing likely

Reply:
I think we need to figure out a different api to do those accesses we do in fil-proofs, that we can support with both backing stores

Reply:
Ok, I'll see what can be done on the `rust-fil-proofs` side to simplify and standardize a bit more how we access the tree.